File: 2477-number-of-ways-to-reach-a-position-after-exactly-k-steps/number-of-ways-to-reach-a-position-after-exactly-k-steps.cpp
class Solution {
public:

    int MOD=1e9+7;
    int dp[3001][3001];
    int end,start;
    int solve(int steps,int position)
    {
        if(position==end && steps==0)
        return 1;

        if(steps==0)
        return 0;

        if(dp[steps][position]!=-1)
        return dp[steps][position];

        int left=solve(steps-1,position-1);

        int right=solve(steps-1,position+1);

        int totalWays=(right%MOD + left%MOD)%MOD;
        return dp[steps][position]=totalWays;
    }

    int numberOfWays(int startPos,int endPos,int k) 
    {
        int offset=1000;
        start=startPos+offset;
        end=endPos+offset;
        memset(dp,-1,sizeof(dp));
        return solve(k,start);
    }
};

File: 2477-number-of-ways-to-reach-a-position-after-exactly-k-steps/number-of-ways-to-reach-a-position-after-exactly-k-steps.java
class Solution {
    private final int mod = 1000000007;
    private int start, target;
    private Integer[][] dp;
    private int solve(int pos, int k)
    {
        if(k == 0)
        return pos == target ? 1 : 0;

        if(dp[pos][k] != null)
        return dp[pos][k];

        int ahead = solve(pos + 1, k - 1) % mod;
        int back = solve(pos - 1, k - 1) % mod;

        return dp[pos][k] = (ahead + back) % mod;
    }

    public int numberOfWays(int startPos, int endPos, int k) 
    {
        start = k + startPos;
        target = k + endPos;

        dp = new Integer[3001][3001];
        return solve(start, k);
    }
}